### PR TITLE
bug 1274413: json_verify.py dnt

### PR DESCRIPTION
**Pull request by groovecoder**

### CODE REVIEW

**Comment by groovecoder** [comment]
Assigning to @fmarier.

File: scripts/json_verify.py [hunk 1]
```diff
@@ -132,10 +132,8 @@ def reset():
 
 
 def start():
-    for root, dirs, files in os.walk("."):
-        for f in files:
-            if ".json" in f:
-                run(f)
+    for f in glob.glob("*.json"):
```
Comment (fmarier) [inline scripts/json_verify.py]: Much nicer!

File: scripts/json_verify.py [hunk 2]
```diff
@@ -100,6 +100,13 @@ def find_uris(categories_json):
             for entity_name, entity_json in entity.iteritems():
                 assert type(entity_name) is UnicodeType
                 assert type(entity_json) is DictType
+                # pop dnt out of the dict, so we can iteritems() over the rest
+                try:
+                    dnt_value = entity_json.pop('dnt', '')
```
Comment (fmarier) [inline scripts/json_verify.py]: I assume that if `dnt` is missing, it returns the empty string here?

**Comment by fmarier** [comment]
r+, but please test a bunch of invalid JSON files in a different repo before merging. I want to make sure that we continue to see Travis errors if a PR breaks the list.

File: scripts/json_verify.py (hunk 2 above)
Comment (groovecoder) [inline scripts/json_verify.py]: Yes, `.pop('dnt', '')` returns the `dnt` item out of `entity_json` or defaults to `''`. So if the `dnt` item is missing, this returns an empty string.

**Comment by groovecoder** [comment]
Yeah, I agree. I don't want to get into a rabbit-hole, but I'm considering turning the script into an actual python test case and creating multiple JSON fixtures with lots of error cases.

**Comment by groovecoder** [comment]
Okay, I only spent a **little** too much time trying to turn this into a python `unittest.TestCase` before I just gave the `json_verify.py` script a filename argument.

I'll start adding some invalid JSON files into a `test/` directory here to make it easy to test the script against invalid files. Do you have any good examples of invalid JSON files? Maybe some that include mistakes we've seen in the past?

**Comment by groovecoder** [comment]
Okay, I added a known-valid and a known-invalid test fixture JSON file. I spot-checked that they pass/fail as expected. If you can spot-check, then I'm ready to merge this if you are.

But I could also look at running the script against all 3 files via tox and setting the known-invalid job as an allowed fail on travis ... ?

File: tests/disconnect-blacklist-valid.json [hunk 3]
```diff
@@ -0,0 +1,121 @@
+{
+    "license": "Copyright 2010-2015 Disconnect, Inc. / This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version. / This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details. / You should have received a copy of the GNU General Public License along with this program. If not, see <http://www.gnu.org/licenses/>.",
+    "categories": {
+        "Disconnect": [
+            {
+                "Facebook": {
+                    "dnt": "dnt",
```
Comment (fmarier) [inline tests/disconnect-blacklist-valid.json]: This should be invalid. The allowed values are: `""`, `"eff"` and `"w3c"`.

**Comment by fmarier** [comment]
Since you have a nice test harness for this now, I would suggest also adding a valid file but with a `"dnt": "bogus"` line (which makes it invalid).

File: tests/disconnect-blacklist-valid.json (hunk 3 above)
Comment (groovecoder) [inline tests/disconnect-blacklist-valid.json]: https://bugzilla.mozilla.org/show_bug.cgi?id=1274413#c0 says "eff", "dnt", or "".

**Comment by groovecoder** [comment]
Added file with a bogus dnt value. Also updated README to include instructions for using the json_verify script and the test fixtures. I also protected the `master` branch here so that only pull requests/commits that have passed the Travis CI job can be merged into master.

**Comment by groovecoder** [comment]
Still need to know if the `""`, `"eff"`, and `"w3c"` options are the valid ones, or if `""`, `"eff"`, and `"dnt"` options from the original bug description are the valid ones.

**Comment by fmarier** [comment]
> Still need to know if the "", "eff", and "w3c" options are the valid ones, or if "", "eff", and "dnt" options from the original bug description are the valid ones.

Yes, that was a typo on bugzilla, sorry about that.

`dnt` is not valid. `eff`, `w3c` and the empty string are the valid ones.

**Comment by groovecoder** [comment]
Updated script and test fixtures with "w3c" instead of "dnt"

File: README.md [hunk 4]
```diff
@@ -1,24 +1,67 @@
...
 
 
-### Raw lists
+## Lists
 These two JSON files power Tracking Protection in Firefox. 
 
-* disconnect-blacklist.json is a version controlled copy of the Disconnect.me public GPL anti-tracking list available here: https://services.disconnect.me/disconnect-plaintext.json
-* disconnect-entitylist.json is a transformed version of the public GPL list that Disconnect reformats for Firefox, in order to reduce the size of the entity list we send to Firefox via the Shavar service. 
+* `disconnect-blacklist.json` is a version controlled copy of the Disconnect.me public GPL anti-tracking list available here: [https://services.disconnect.me/disconnect-plaintext.json](https://services.disconnect.me/disconnect-plaintext.json)
```
Comment (fmarier) [inline README.md]: You can also use `<http://example.com>` in markdown linkify URLs without having to repeat them.

**Comment by fmarier** [comment]
r+

**Pull request merged**